### ISSUE TITLE
[Apline] Need a better approach to fix CMSG_NXTHDR on Alpine Linux (musl-libc)

### ISSUE THREAD

**Issue by ghost** [dotnet/corefx#6312]
The current master ([`5b793bc`](https://github.com/dotnet/corefx/tree/5b793bc735d6a63b7f4be466df3351008b27a426)) breaks the native build (`corefx/build.sh native`) at 70% on Alpine Linux with the following error:

``` ash
/root/corefx/src/Native/System.Native/pal_networking.cpp:1311:31: error: cast from 'unsigned char *' to 'struct cmsghdr *'
increases
      required alignment from 1 to 4 [-Werror,-Wcast-align]
             controlMessage = CMSG_NXTHDR(&header, controlMessage))
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/sys/socket.h:267:10: note: expanded from macro 'CMSG_NXTHDR'
        ((struct cmsghdr *)__CMSG_NEXT(cmsg))))
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Code comparison between glibc and musl-libc:
- musl-libc: https://fossies.org/dox/musl-1.1.14/include_2sys_2socket_8h.html#a4f70853d9633fb80428189cd7bb292d0 (click on `__CMSG_NEXT` in that macro to see `unsigned char*` cast)
- glibc: http://lxr.free-electrons.com/source/include/linux/socket.h#L89 (click on `__cmsg_nxthdr` in the called macro to notice that it returns `cmsghdr *`)

https://github.com/dotnet/corefx/pull/6263 provides a workaround to turn that warning off inline for this particular call to macro and then restores it using Clang's `pragma`tic directives. The warning disabling only affects the musl-libc systems (enclosed by `#ifndef __GLIBC__`).

This probably needs to be fixed upstream.

cc @pgavlin, @benpye, @ncopa

**Comment by karelz** [comment]
Old issue. Most likely not a problem anymore. If it is, feel free to reopen.

**Comment by kasper3** [comment]
This is still unresolved what it was tracking: https://github.com/dotnet/corefx/blob/57ff88bb75a0/src/Native/Unix/System.Native/pal_networking.c#L811-L829.

**Comment by kasper3** [comment]
And removing this patch still gives error on Alpine Linux v3.6. This should be reopened. cc @janvorli, @stephentoub

**Comment by janvorli** [comment]
@kasper3 I've looked into it today. The reason why the compiler gives the warning is not due to the compiler not knowing that the /usr/include/sys/socket.h. The problem is that the incorrect code that it warns about comes from a macro expansion. So any warnings in the header itself would be suppressed, but not warnings from expanded macros. There are various notes about this on the internet. 
glibc solves this by creating an inline wrapper in one of the glibc headers. Then clang can suppress it as that header is considered a system one.
GCC is obviously better in detecting this case, since it doesn't warn (I've tried with a little test app).
So it seems that keeping the pragma to turn of the warning for this macro invocation is a reasonable way. We could also create an extra header containing a function wrapping the macro invocation and marking that header as a system one (clang has a pragma for that), but that seems too clunky.
